### PR TITLE
Base BreachAlertsManager class with loadBreaches() + compare()

### PR DESCRIPTION
**Pull request by vphong**
#6644

### CODE REVIEW

File: Client/Frontend/Login Management/BreachAlertsManager.swift [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            if let decoded = try? decoder.decode([BreachRecord].self, from: data) {
+                DispatchQueue.main.async {
+                    self.breaches = decoded
+                }
+            }
+        }
+
+        dataTask?.resume()
+
+    }
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsManager.swift:93]: I'm wondering if we should separate the actual data request out into a client that just handles doing that work? Call it BreachAlertsClient or something like that? Looking around the app I see that as a pattern for data tasks.

File: Client/Frontend/Login Management/BreachAlertsManager.swift [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    /// Compares a list of logins to a list of breaches and returns breached logins.
+    ///    - Parameters:
+    ///         - logins: a list of logins to compare breaches to
+    func compareToBreaches(_ logins: [Any])  {
+
+        for entry in logins {
+            let login = entry as! LoginRecord; // typecast for ease of use
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsManager.swift:101]: We should remove this force unwrapping and I'm thinking logins should probably have a type instead of `Any`?

File: Client/Frontend/Login Management/BreachAlertsManager.swift [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            guard let data = data, !data.isEmpty else {
+                print("loadBreaches(): invalid data")
+                return
+            }
+
+            let decoder = JSONDecoder()
+
+            // .convertFromPascalCase
+            decoder.keyDecodingStrategy = .custom { keys in
+                let key = keys.last! // make sure array not empty
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsManager.swift:72]: We should remove this force unwrapping as well

File: Client/Frontend/Login Management/BreachAlertsManager.swift [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+                // string manipulation
+                let keyStr = key.stringValue
+                let pascalToCamel = keyStr.prefix(1).lowercased() + keyStr.dropFirst()
+
+                // CodingKey conformity
+                let codingKeyType = type(of: key)
+                return codingKeyType.init(stringValue: pascalToCamel)!
+
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsManager.swift:81]: You can set custom keys for the codable variables if their names don't match to the keys in the data you get back. I think that would be ideal here.

File: Client/Frontend/Login Management/BreachAlertsManager.swift [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                if loginHostURL?.baseDomain == breach.domain {
+                    print("compareToBreaches(): breach: \(breach.domain)")
+                }
+
+                // date check
+                let pwLastChanged = Date.init(timeIntervalSince1970: TimeInterval(login.timePasswordChanged))
+
+                let dateFormatter = DateFormatter()
+                dateFormatter.dateFormat = "yyyy-MM-dd"
+                let breachDate = dateFormatter.date(from: breach.breachDate)!
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsManager.swift:115]: Another force unwrapping mention, we shouldn't have any of that in the final version of this PR

File: Client/Frontend/Login Management/LoginListViewController.swift [hunk 6]
```diff
@@ -250,6 +251,12 @@ class LoginListViewController: SensitiveViewController {
         let deferred = Deferred<Maybe<[LoginRecord]>>()
         profile.logins.searchLoginsWithQuery(query) >>== { logins in
             deferred.fillIfUnfilled(Maybe(success: logins.asArray()))
+
+            // consider taking logins to breachalertsmanager from here
+            // which indexes need a special layout? to display breached login
+            breachAlertsManager.loadBreaches()
+            breachAlertsManager.compareToBreaches(logins.asArray())
+
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/LoginListViewController.swift:259]: Maybe remove this portion for now, as I think handling the breach alerts result is a separate chunk of work for it's own PR and it's not fully finished here.

File: Client/Frontend/Login Management/BreachAlertsManager.swift (hunk 2 above)
Comment (vphong) [reply Client/Frontend/Login Management/BreachAlertsManager.swift:101]: I was using `Any` for portability, but I suppose if this file were to be used in another project, the parameter type itself could just be changed. Changing this now

File: Client/Frontend/Login Management/BreachAlertsManager.swift (hunk 4 above)
Comment (vphong) [reply Client/Frontend/Login Management/BreachAlertsManager.swift:81]: All of the endpoint keys are in Pascal case, so I was thinking along the lines of using the Swift-given `decoder.keyDecodingStrategy = .convertFromSnakeCase` except there is no .convertFromPascalCase so I made a custom one. Maybe this block of code is better suited to an extension though

File: Client/Frontend/Login Management/BreachAlertsManager.swift (hunk 1 above)
Comment (vphong) [reply Client/Frontend/Login Management/BreachAlertsManager.swift:91]: When I study the cases with the other clients used in the app, they either seem to be 1) whatever is calling the client is complex or 2) the network requests are many/complex. I think in this case, keeping this network call just confined to this class should be okay, since it's pretty simple and the max it will do is change the UI.

File: Client/Frontend/Login Management/BreachAlertsManager.swift [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    var pwnCount: Int
+    var description: String
+    var logoPath: String
+    var isVerified: Bool
+    var isFabricated: Bool
+    var isSensitive: Bool
+    var isRetired: Bool
+    var isSpamList: Bool
+    var logoUrl: String?
+}
```
[the remark this answers is not in the document]
Comment (vphong) [reply Client/Frontend/Login Management/BreachAlertsManager.swift:24]: We can definitely cull some. I know which ones we need right now but I'm not sure which ones we might need in the future

File: Client/Frontend/Login Management/BreachAlertsManager.swift (hunk 1 above)
Comment (kaylagalway) [reply Client/Frontend/Login Management/BreachAlertsManager.swift:91]: Chatted outside of github about this and we came to the conclusion that a Client will be added for better testability and future expansion of breach alerts requests, if ever necessary.

File: Client/Frontend/Login Management/BreachAlertsClient.swift [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import Foundation
+import Shared
+
+/// Errors related to BreachAlertsClient and BreachAlertsManager.
+struct BreachAlertsError: MaybeErrorType {
+    public let description: String
+    public init(description: String) {
+        self.description = description
+    }
+}
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsClient.swift:18]: Is this init required for the `MaybeErrorType` protocol conformance? If not, I don't believe you need it as structs in Swift come with initializers baked in for their variables.

File: Client/Frontend/Login Management/BreachAlertsClient.swift [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+/// Handles all network requests for BreachAlertsManager.
+public class BreachAlertsClient: BreachAlertsClientProtocol {
+    private var dataTask: URLSessionDataTask?
+    public enum Endpoint: String {
+        case breachedAccounts = "https://monitor.firefox.com/hibp/breaches"
+    }
+
+    public init() { }
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsClient.swift:31]: This init can be removed as well since it doesn't take any parameters

File: Client/Frontend/Login Management/BreachAlertsClient.swift [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+            guard let data = data, !data.isEmpty else {
+                completion(Maybe(failure: BreachAlertsError(description: "empty data")))
+                return
+            }
+
+            completion(Maybe(success: data))
+
+        }
+
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsClient.swift:60]: The extra whitespaces in this function between if and guard blocks, and after the completion, can all be removed

File: Client/Frontend/Login Management/BreachAlertsClient.swift [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        dataTask?.resume()
+    }
+
+
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsClient.swift:64]: nit: extra whitespace here as well, only one line of whitespace necessary between functions

File: Client/Frontend/Login Management/BreachAlertsClient.swift [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+                if let responseType = response.allHeaderFields["Content-Type"] as? String {
+                    return responseType.contains(type) ? response : nil
+                }
+                return nil
+            }
+            return response
+        }
+        return nil
+    }
+}
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsClient.swift:81]: Does this code need to be copied here? Or can you just use the function that is already in NetworkUtils?

File: Client/Frontend/Login Management/BreachAlertsManager.swift [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /// Loads breaches from Monitor endpoint using BreachAlertsClient.
+    ///    - Parameters:
+    ///         - completion: a completion handler for the processed breaches
+    func loadBreaches(completion: @escaping (Maybe<[BreachRecord]>) -> Void) {
+        print("loadBreaches(): called")
+
+        DispatchQueue.global(qos: .background).async {
+            self.breachAlertsClient.fetchData(endpoint: .breachedAccounts) { maybeData in
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/BreachAlertsManager.swift:46]: You don't need to manually set this to a background thread, since it has a completion handler and data tasks run asynchronously, it will already be on a background asynchronous thread.

File: Client/Frontend/Login Management/LoginListViewController.swift [hunk 14]
```diff
@@ -250,6 +251,7 @@ class LoginListViewController: SensitiveViewController {
         let deferred = Deferred<Maybe<[LoginRecord]>>()
         profile.logins.searchLoginsWithQuery(query) >>== { logins in
             deferred.fillIfUnfilled(Maybe(success: logins.asArray()))
+
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/LoginListViewController.swift:254]: nit: remove this whitespace between deferred and succeed

File: Client/Frontend/Login Management/LoginListViewController.swift [hunk 15]
```diff
@@ -50,6 +50,7 @@ class LoginListViewController: SensitiveViewController {
     fileprivate var selectionButtonHeightConstraint: Constraint?
     fileprivate var selectedIndexPaths = [IndexPath]()
     fileprivate let tableView = UITableView()
+    fileprivate let breachAlertsManager = BreachAlertsManager()
```
Comment (kaylagalway) [inline Client/Frontend/Login Management/LoginListViewController.swift:53]: This `breachAlertsManager` can be removed from here for now as we aren't using it yet.

File: Client/Frontend/Login Management/BreachAlertsClient.swift (hunk 8 above)
Comment (vphong) [reply Client/Frontend/Login Management/BreachAlertsClient.swift:18]: I was following LoginRecordError for this https://github.com/mozilla-mobile/firefox-ios/blob/4436268ba1c5fb3beea390e0b9b72b2c88a95fc8/Storage/Rust/RustLogins.swift#L78-L83
but now that I look around the codebase, there doesn't seem to be a standard for defining `MaybeErrorType`s. I took out the init and it compiles fine so I'll add this in the next commit

File: Client/Frontend/Login Management/BreachAlertsClient.swift (hunk 9 above)
Comment (vphong) [reply Client/Frontend/Login Management/BreachAlertsClient.swift:31]: The compiler was upset with me for a bit, saying `BreachAlertsClient` was internal, so I defined a public init and it was happy. Fiddling final/public left this init behind, and it compiles fine without this public init now so I'll remove it

File: Client/Frontend/Login Management/BreachAlertsClient.swift (hunk 12 above)
Comment (vphong) [reply Client/Frontend/Login Management/BreachAlertsClient.swift:81]: Remnant of when I was trying not to import stuff. But now that we've imported `Shared` for `Maybe` it works just fine from NetworkUtils.

Comment (kaylagalway) [reply Client/Frontend/Login Management/BreachAlertsClient.swift:81]: Sweet
